### PR TITLE
Add local methods

### LINKED ISSUE #32: Local Method Support
We need to be able to allow end developers to add local methods in their blueprints.

For example: 

```ruby
class UserBlueprint < Blueprinter::Base
  field :full_name

  def full_name
    "#{object.first_name} #{object.last_name}"
  end
end
```

### PR DESCRIPTION
**Pull request by philipqnguyen**
Resolves #32

~Add a method call `local_method`.  `local_method` accepts a method name and a method body. The method name is passed as an argument, and the method body is passed into a block.~

~Why we need something like this? There are situations where end devs want to define a method that are not used anywhere else except to have it serialized into JSON.~

Usage:
```ruby
class UserBlueprint < Blueprinter::Base
  identifier :id
  fields :name
  local_method :age { 30 + 1 }
end
```

~So currently this works. But I also just now realized that it's use cases are currently narrow without first tackling #33. #33 would allow end developers to pass in custom options from `render` and utilize that in their blueprint class.~

Update: 10/20
~Now object is passed into the block so you can use it like so:~
``` ruby
class UserBlueprint < Blueprinter::Base
  identifier :id
  fields :first_name
  local_method(:full_name) do |obj|
    "#{obj.first_name} #{obj.last_name}"
  end
end
```

Update 10/23
We allow the passing of a block to `field` like so:
```ruby
class UserBlueprint < Blueprinter::Base
  # This would create a key value pair in the json output that looks like this:
  # {"full_name": "Jon Doe"}
  field(:full_name) do |obj|
    "#{obj.first_name} #{obj.last_name}"
  end
end
```

### CODE REVIEW

File: lib/blueprinter/serializers/local_method_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class Blueprinter::LocalMethodSerializer < Blueprinter::Serializer
+  def serialize(field_name, _object, options = {})
+    options[:local_methods][field_name]
```
Comment (dlcarter) [inline lib/blueprinter/serializers/local_method_serializer.rb]: If we change to this:
`options[:local_methods][field_name].call(_object, options)`

We should get the ability to interact with both the object AND any options passed into the blueprint, right?

Comment (philipqnguyen) [reply lib/blueprinter/serializers/local_method_serializer.rb]: that's a great idea

Comment (philipqnguyen) [reply lib/blueprinter/serializers/local_method_serializer.rb]: @dlcarter What do you see as useful for exposing the `options` variable through the block right now?

Right now, `render` does not accept an option args. 

We could allow `local_method` to accept `options` args, but do you see that as being useful right now? Or should that wait for #33 when we work on exposing options arguments for `render`?

**Comment by philipqnguyen** [comment]
Just pushed a commit that would pass the object into the block. So can do this:

```ruby
      local_method(:full_name) do |obj|
        "#{obj.first_name} #{obj.last_name}"
      end
```

**Comment by philipqnguyen** [comment]
We also should discuss more about this. Whether there is an alternative to calling it `local_method`, or if there is a better way all together.

**Comment by AllPurposeName** [comment]
All we're attempting to do is define a field in a different way, so instead of the implied mapping between the field and the object, we're just intercepting and defining a field explicitly.

```ruby
fields :first_name, :last_name
field(:full_name) do |obj|
  "#{obj.first_name} #{obj.last_name}"
end
field(:age) { 30 + 1 }
```

**Comment by philipqnguyen** [comment]
Really like that idea @AllPurposeName, I'll go ahead and update this pr with it.

File: lib/blueprinter/base.rb [hunk 2]
```diff
@@ -49,18 +50,30 @@ def self.identifier(method, name: method, serializer: PublicSendSerializer)
...
+    #   end
+    #
     # @return [Field] A Field object
-    def self.field(method, options = {})
-      name = options.delete(:name) || method
-      serializer = options.delete(:serializer) || AssociationSerializer
-      current_view << Field.new(method, name, serializer, options)
+    def self.field(method, options = {}, &block)
+      options = if block_given?
+        {name: method, serializer: BlockSerializer, block: {method => block}}
```
Comment (AllPurposeName) [inline lib/blueprinter/base.rb]: NB: I suspect a FieldOptions class is yearning to be born.

File: lib/blueprinter/base.rb [hunk 3]
```diff
@@ -49,18 +50,30 @@ def self.identifier(method, name: method, serializer: PublicSendSerializer)
...
+      else
+        {name: method, serializer: AssociationSerializer}
+      end.merge options
```
Comment (AllPurposeName) [inline lib/blueprinter/base.rb]: Would you mind putting parens here? 😛

File: lib/blueprinter/base.rb [hunk 4]
```diff
@@ -49,18 +50,30 @@ def self.identifier(method, name: method, serializer: PublicSendSerializer)
...
+      current_view << Field.new(method, options[:name], options[:serializer],
+        options)
```
Comment (dlcarter) [inline lib/blueprinter/base.rb]: NB: The `options)` bugs me - would love to clean up the formatting of this line

File: lib/blueprinter/base.rb (hunk 3 above)
Comment (philipqnguyen) [reply lib/blueprinter/base.rb]: ohhkaayy 😄

File: lib/blueprinter/base.rb (hunk 4 above)
Comment (philipqnguyen) [reply lib/blueprinter/base.rb]: will do

**Comment by trivedi** [comment]
@philipqnguyen is there any way to call a local method without defining a block? something similar to

```
field :age

def age
  18
end
```


the reason I ask is because sometimes the block can be large, so it's nice to break it into its own method. but a method that is only used in the serializer and not in the model (as mentioned in your PR summary).

**Comment by philipqnguyen** [comment]
@trivedi

You can do 
```ruby
field :age do
  age
end

def self.age
  18
end
```

**Pull request merged**